qiskit/backends/ibmq/ibmqjob.py

```diff
@@ -445,61 +422,53 @@ def _wait_for_job(self, timeout=60, wait=5):
-                self._exception = QISKitError("get_job didn't return status: %s" %
-                                              (pprint.pformat(api_result)))
-                raise QISKitError("get_job didn't return status: %s" %
-                                  (pprint.pformat(api_result)))
 
+            logger.info('status = %s (%d seconds)', self._status, elapsed_time)
             time.sleep(wait)
-            api_result = self._update_status()
 
         if self.cancelled:
-            job_result = {'id': self._id, 'status': 'CANCELLED',
-                          'result': 'job cancelled'}
-            return Result(job_result)
-
-        elif self.exception:
-            job_result = {'id': self._id, 'status': 'ERROR',
-                          'result': str(self.exception)}
-            return Result(job_result)
-
-        if api_result is None:
-            api_result = self._api.get_job(self._id)
+            return Result({'id': self._id, 'status': 'CANCELLED',
+                           'result': 'job cancelled'})
 
+        job_data = self._api.get_job(self._id)
         job_result_list = []
-        for circuit_result in api_result['qasms']:
+        for circuit_result in job_data['qasms']:
             this_result = {'data': circuit_result['data'],
                            'name': circuit_result.get('name'),
                            'compiled_circuit_qasm': circuit_result.get('qasm'),
                            'status': circuit_result['status']}
             if 'metadata' in circuit_result:
                 this_result['metadata'] = circuit_result['metadata']
+
             job_result_list.append(this_result)
-        job_result = {'id': self._id,
-                      'status': api_result['status'],
-                      'used_credits': api_result.get('usedCredits'),
-                      'result': job_result_list}
-        job_result['backend_name'] = self.backend_name
-        return Result(job_result)
 
+        return Result({'id': self._id,
+                       'status': job_data['status'],
+                       'used_credits': job_data.get('usedCredits'),
+                       'result': job_result_list,
+                       'backend_name': self.backend_name})
+
+    def _wait_for_submitting(self):
+        """ Waits for the request to return a Job ID """
+        if self._id is None:
+            if self._future is None:
+                raise JobError("You have to submit before asking for status or results!")
+            submit_info = self._future.result(timeout=60)
+            if 'error' in submit_info:
+                self._status = JobStatus.ERROR
+                raise JobError(str(submit_info['error']))
 
-class IBMQJobError(QISKitError):
-    """class for IBM Q Job errors"""
-    pass
+            self._creation_date = submit_info.get('creationDate')
+            self._status = JobStatus.QUEUED
+            self._id = submit_info.get('id')
 
 
 def _reorder_bits(result):
```

Comment:
Not technically part of your changes, and more as a "wishlist" item: can you check line 511 in this function:
```
raise ResultError("creg sizes don't add up in result header.")
```
It seems `ResultError` [expects a dict](https://github.com/Qiskit/qiskit-terra/blob/master/qiskit/_resulterror.py#L29) instead of a string: maybe it would make sense either to throw a `JobError` if it makes sense conceptually, or to revise the argument?